walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,68 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
 
 public final class FindMeetingQuery {
+  /**
+   *  Solves for the possible time ranges available to schedule a meeting
+   *  with the given attendees and current events going on.
+   *
+   *  @param events Collection of events that are going on during that day
+   *  @param request A requested meeting with certain attendees
+   *  
+   *  @return A collection of time ranges possible to schedule
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<TimeRange> eventRanges = sortedRangesInQuestion(events, request.getAttendees());
+    List<TimeRange> possibleRanges = new ArrayList<TimeRange>();
+    if (eventRanges.size() >= 1){
+      TimeRange startOfDay = TimeRange.fromStartEnd(TimeRange.START_OF_DAY, eventRanges.get(0).start(), false);
+      if (startOfDay.duration() >= request.getDuration()){
+        possibleRanges.add(startOfDay);
+      }
+    }
+    int startingPointer = 0;
+    for (int endingPointer = 1; endingPointer < eventRanges.size(); endingPointer++) {
+      if (eventRanges.get(startingPointer).overlaps(eventRanges.get(endingPointer))) {
+        startingPointer = endingPointer;
+      }
+      else {
+        TimeRange between = TimeRange.fromStartEnd(eventRanges.get(startingPointer).end(),
+            eventRanges.get(endingPointer).start(), false);
+        if (between.duration() >= request.getDuration()){
+          possibleRanges.add(between);
+        }
+        startingPointer++;
+      }
+    }
+    if (eventRanges.size() >= 1){
```

Comment:
This if statement is repeated above - do you need both instances?

Reply:
I repeat this statement below so that the ending time range is at the end of possibleRanges, I suppose I could add this function above and then simply sort possibleRanges

Reply:
see: carolyn's comment above https://github.com/brettallenyo/step2020/pull/17#discussion_r441863790

Reply:
I still don't see why you need 2 if statements, esp when the conditions are the same. Now you're doing 
```
if (eventRanges.size()>= 1){
 // Do something
}
if (eventRanges.size()>= 1){
 // Do something else
}
```

At the very least you can combine both if statements and do: (this is independent of whether the code logic is correct)
```
if (eventRanges.size()>= 1){
 // Do something
 // Do something else
}
```

Reply:
I moved the code because carolyn pointed out that I only have to sort the time ranges because i add the beginning and ending ranges at the very start, then add the rest of the time ranges in order. So in reality, I sort the entire List just to get one value to the end. To reduce computation time, I could just move adding the ending time range to the end and repeating the if-statement. Nevertheless, I removed it, moved it up into the other if statement, and re-added the Collections.sort operation at the end.